core/src/main/scala/org/apache/spark/api/python/VirtualEnvFactory.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.api.python
+
+import java.io.File
+import java.util.{Map => JMap}
+import java.util.Arrays
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.collection.JavaConverters._
+
+import com.google.common.io.Files
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+
+class VirtualEnvFactory(pythonExec: String, conf: SparkConf, isDriver: Boolean)
+  extends Logging {
+
+  private val virtualEnvType = conf.get("spark.pyspark.virtualenv.type", "native")
+  private val virtualEnvBinPath = conf.get("spark.pyspark.virtualenv.bin.path", "")
+  private val initPythonPackages = conf.getOption("spark.pyspark.virtualenv.packages")
```

Comment:
So if the factory is made once then how will these get updated?

Reply:
For the existing running executor,  the only way to install additional packages is via `sc.intall_packages`.  And `spark.pyspark.virtualenv.packages` will be updated on driver side first when `sc.install_packges` is called, then new allocated executor will fetch this property and install all the additional packages properly.